### ISSUE TITLE
drop support for Ruby 3.0, add support for Ruby 3.4

### ISSUE THREAD

**Issue by flavorjones** [sparklemotion/nokogiri#3273]
Ruby 3.0 was EOL 23 Apr 2024. Nokogiri's policy is to support the current and previous three versions of Ruby, so we won't drop support until Ruby 3.4 is released.

Tagging this for v1.18.0 which will coincide with Ruby 3.4.0 final release in Dec 2024.

This will likely happen in v1.18 release candidates earlier than December.

- [x] drop 3.0 from cross-rubies and CI and docker.rake
- [x] add 3.4 to cross-rubies and docker.rake
- [x] delete 3.0 valgrind suppressions
- [x] remove windows MSVCRT builds
- [x] update gemspec `required_ruby_version` to `">= 3.1.0"`
- [x] update README
- [x] bump rubocop `TargetRubyVersion` to `"3.1"`
- [x] remove any tests or code relevant only to Ruby 3.0
- [x] remove unnecessary extconf.rb compile-time checks

**Comment by koic** [comment]
This is just one perspective I’d like to share. When I tried using the preview version of Ruby, I encountered a build error in the Linux distribution:

```console
bundle install
(snip)

Fetching gem metadata from https://rubygems.org/..........
nokogiri-1.17.1-x86_64-linux requires ruby version >= 3.0, < 3.4.dev, which is incompatible with the current version, 3.4.0.preview2
```

Nokogiri holds an important position in the Ruby ecosystem, and it would be beneficial if installation errors due to required Ruby versions did not occur with the preview version. Early adopters aiming to detect issues early would likely want to start testing the behavior after successfully installing Nokogiri as soon as possible. Thank you.

**Comment by flavorjones** [comment]
@koic This is a long-known limitation in bundler and not an issue with Nokogiri. You should be able to delete your `Gemfile.lock` and re-bundle (it should download `nokogiri-1.17.1.gem` and compile from source during installation).

**Comment by flavorjones** [comment]
See https://github.com/rubygems/rubygems/issues/6237 for an example of a similar bug report to bundler.

**Comment by flavorjones** [comment]
Just to clarify: "add support for Ruby 3.4" in this issue means "add precompiled libraries for Ruby 3.4 to the native gems". Ruby 3.4 is already supported, we run CI with ruby-head. But you must install the non-native gem.

**Comment by koic** [comment]
I see! Thank you for the clear explanation, and for always keeping things well-maintained!

**Comment by flavorjones** [comment]
This issue will be closed by #3376

**Comment by flavorjones** [comment]
I've released v1.18.0.rc1 with native gems containing Ruby 3.4-compatible libraries.

https://github.com/sparklemotion/nokogiri/releases/tag/v1.18.0.rc1
